text/3806-align-attr.md

```diff
@@ -0,0 +1,380 @@
+the containing type, the attribute has no effect.
+
+In contrast to a `repr(align(…))` wrapper struct, an `align` annotation does *not*
+necessarily add extra padding to force the field to have a size that is a
+multiple of its alignment. (The size of the containing ADT must still be a
+multiple of its alignment; that hasn't changed.)
+
+`align` attributes for fields of a `#[repr(packed(n))]` ADT may not specify an
+alignment higher than `n`.
+
+```rust
+#[repr(packed(4))]
+struct Sardines {
+    #[align(2)] // OK
+    a: u8,
+    #[align(4)] // OK
+    b: u16,
+    #[align(8)] //~ ERROR
+    c: u32,
+}
+```
+
+`align()` attributes on ADT fields are shown in `rustdoc`-generated documentation.
+
+## Interaction with `repr(C)`
+
+`repr(C)` currently has two contradictory meanings: “a simple, linear layout
+algorithm that works the same everywhere” and “an ABI matching that of the
+target’s standard C compiler”. This RFC does not aim to reslove that conflict;
+that is being discussed as part of [RFC
+3718](https://github.com/rust-lang/rfcs/pull/3718). Henceforth, we will use
+`repr(C_for_real)` to denote “match the system C compiler”, and `repr(linear)`
+to denote “simple, portable layout algorithm”; but those names are not
+normative.
+
+### `repr(C_for_real)`
+
+The layout of a `repr(C_for_real)` ADT with `align` attributes on its fields is
+identical to that of the corresponding C ADT declared with `alignas`
+annotations. For example, the struct below is equivalent to the C `struct foo`
+from the motivation section:
+
+```rust
+#[repr(C_for_real)]
+pub struct foo {
+    pub x: u8,
+    #[align(128)]
+    pub y: u8,
+    pub z: u8,
+}
+```
+
+### `repr(linear)`
+
+In a `repr(linear)` ADT, a field with an `align` attribute has its alignment, as
+well as the alignment of the containing ADT, increased to at least what the
+attribute specifies.
+
+For example, the following two structs have the same layout in memory (though
+not necessarily the same ABI):
+
+```rust
+#[repr(linear)]
+pub struct foo {
+    pub x: u8,
+    #[align(128)]
+    pub y: u8,
+    pub z: u8,
+}
+```
+
+```rust
+#[repr(linear, align(128))]
+pub struct foo2 {
+    pub x: u8,
+    pub _padding: [MaybeUninit<u8>; 127usize],
+    pub y: u8,
+    pub z: u8,
+}
+```
+
+## On `static`s
+
+Any `static` item (including `static`s inside `extern` blocks) may have an
+`align` attribute applied:
+
+```rust
+#[align(32)]
+static BAZ: [u32; 12] = [0xDEADBEEF; 12];
+
+unsafe extern "C" {
+    #[align(2)]
+    safe static BOZZLE: u8;
+}
+```
+
+The effect of the attribute is to force the `static` to be stored with at least
+the specified alignment. The attribute does not force padding bytes to be added
+after the `static`. For `static`s inside `unsafe extern` blocks, if the `static`
+does not meet the specified alignment, the behavior is undefined. (For
+misaligned `static` items declared inside old-style `extern` blocks, UB occurs
+only if the item is used.)
```

Comment:
Why make this distinction? Making UB depend on surface syntax for essentially the same thing might UB detection (e.g., miri or rustc-inserted "UB checks") and exploitation more difficult, and since this is a new feature there's not even backwards compatibility concerns.

Reply:
Making it UB even if not accessed enables more optimizations. For example, the compiler could hoist an access to the `static` outside of an `if`. And we can always relax this UB in the future, if it's not worth it

Reply:
I don’t object to having UB if the item isn’t accessed. I just think it should be the same rules for `unsafe extern {}` and `extern {}` so later stages don’t have to carry around the distinction.

Reply:
But old-style `extern` is safe, and safe code should not be able to cause UB.

Reply:
@hanna-kruppe is correct; it'd be better to simply require `unsafe extern` in order to use `#[align(..)]`.

Reply:
`#[align]` isn't special. The UB that can result if an `extern` static item specifies an incorrect alignment, is the same as the UB if such an item specifies an incorrect type. So restricting this feature to edition ≥2024 would not accomplish anything.

Reply:
> So restricting this feature to edition ≥2024 would not accomplish anything.

`unsafe extern` can be used in any edition.

Reply:
Restricting to `unsafe extern`, then.

Reply:
I’ve updated the RFC to make it clear that this case of UB should obey the same rules as UB from the static’s value not being valid for its type, whatever those rules are.